### PR TITLE
Add e2e disprove tests for positive and negative scenario

### PR DESCRIPTION
**Pull request by StarDumpling**

### CODE REVIEW

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+use tokio::time::sleep;
+
+use crate::bridge::{
+    faucet::{Faucet, FaucetType},
+    helper::{generate_stub_outpoint, get_default_peg_out_event, TX_WAIT_TIME},
+    setup::{setup_test, INITIAL_AMOUNT},
+};
+
+pub async fn create_peg_in_graph(
+    verifier_0: &mut BitVMClient,
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:26]: We're passing two verifiers and no operator. I know this is just variable naming, but to avoid confusion I suggest we name the one that creates the graph an 'operator'.

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+use std::time::Duration;
+use tokio::time::sleep;
+
+use crate::bridge::{
+    faucet::{Faucet, FaucetType},
+    helper::{generate_stub_outpoint, get_default_peg_out_event, TX_WAIT_TIME},
+    setup::{setup_test, INITIAL_AMOUNT},
+};
+
+pub async fn create_peg_in_graph(
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:25]: The `create_peg_in/out_graph()` funcs are already defined in a couple other places in test code: `musig2_peg_out.rs` and `e2e_l2_ui.rs`. Please consider using one of those, if possible.

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    match verifier_0.broadcast_peg_in_deposit(&graph_id).await {
+        Ok(txid) => eprintln!(
+            "Broadcasted {} with txid {txid}",
+            "peg-in deposit".bold().green()
+        ),
+        Err(e) => panic!("Failed to broadcast peg-in deposit: {e}"),
+    }
+    // Wait for peg-in deposit transaction to be mined
+    println!("Waiting for peg-in deposit tx...");
+    sleep(Duration::from_secs(TX_WAIT_TIME)).await;
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:45]: Please use `wait_for_confirmation_with_message()` instead (here copied from `musig2_peg_in.rs:100` in the `rc` branch):
```
    wait_for_confirmation_with_message(config.network, Some("peg-in deposit tx")).await;
```

BTW, please merge the latest `rc` branch, because the file mentioned above is different there. Strangely the PR is not showing it.

Also, you can check out how similar functions have been updated in `rc` (e.g. they use information about the network used in the test to know how long to wait for tx confirmation).

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+    match operator.broadcast_kick_off_1(peg_out_graph_id).await {
+        Ok(txid) => println!(
+            "Broadcasted {} with txid {txid}",
+            "kick-off 1".bold().green()
+        ),
+        Err(e) => panic!("Failed to broadcast kick-off 1: {e}"),
+    }
+    println!("Waiting for kick-off 1 tx...");
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:290]: Please merge `rc`. We use `wait_for_timelock_expiry()` there. The `TX_WAIT_TIME` const has been removed.

**Review by franklynot — CHANGES_REQUESTED** [review]

File: bridge/tests/bridge/helper.rs [hunk 5]
```diff
@@ -282,3 +333,50 @@ pub fn get_incorrect_proof() -> RawProof {
 
     correct_proof
 }
+
+// TODO: Consider importing `gen_correct_proof` fn from bitvm/src/chunker/disprove_execution.rs
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/helper.rs:337]: How it would create a circular dependency? It's like importing `disprove_exec` from the same file that we use in connectorC to generate disprove witness. 

This is also just a test function in both, bitvm and bridge crate. The reason for this it to avoid copying code instead of reusing it. It BitVM team decides to change this function, we would have to change it as well.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 1 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs:26]: Technically, `depositor` is the one that creates peg in graph and we need both verifiers to complete the signing process. 

In this test scenario, verifier_0 plays also depositor role.

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+    verifier_1: &mut BitVMClient,
+    deposit_input: Input,
+    depositor_evm_address: &String,
+) -> String {
+    println!("{}", "Creating PEG-IN graph...".bold().yellow());
+    let graph_id = verifier_0
+        .create_peg_in_graph(deposit_input, depositor_evm_address)
+        .await;
+
+    match verifier_0.broadcast_peg_in_deposit(&graph_id).await {
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs:36]: It would be fewer lines, but I wanted to show every step that takes place during the peg-in and peg-out process.

File: bridge/src/proof.rs [hunk 7]
```diff
@@ -0,0 +1,4 @@
+use bitvm::chunker::disprove_execution::RawProof;
+
+// TODO: replace with actual implementation
+pub fn generate_proof() -> RawProof { RawProof::default() }
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/src/proof.rs]: Renamed to `get_proof`.

File: bridge/src/mod.rs [hunk 8]
```diff
@@ -3,6 +3,7 @@ pub mod connectors;
 pub mod constants;
 pub mod contexts;
 pub mod graphs;
+pub mod proof;
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/src/mod.rs]: removed

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+        Ok(txid) => eprintln!(
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Fixed.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 3 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Done

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 4 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Done

File: bridge/tests/bridge/helper.rs [hunk 10]
```diff
@@ -270,9 +304,26 @@ pub fn get_lock_scripts_cached(
 }
 
 pub fn get_correct_proof() -> RawProof {
-    let correct_proof = RawProof::default();
-    assert!(correct_proof.valid_proof());
-    correct_proof
+    type E = Bn254;
+    let k = 6;
+    let mut rng = ark_std::rand::rngs::StdRng::seed_from_u64(test_rng().next_u64());
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/helper.rs:341]: As far as I can tell, this function should be deterministic. According to comments in `test_rng()` it should return the same value with high probability if `getrandom` feature is not enabled (we don't enable this feature).

File: bridge/tests/bridge/setup.rs [hunk 11]
```diff
@@ -203,6 +203,8 @@ pub async fn setup_test() -> SetupConfig {
     let withdrawer_context =
         WithdrawerContext::new(source_network, WITHDRAWER_SECRET, &n_of_n_public_keys);
 
+    let veryfying_key = get_correct_proof().vk;
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/setup.rs:206]: If `get_correct_proof` is deterministic, we can just use it on the fly, whenever we need it.

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+    let peg_out_confirm_input_amount = Amount::from_sat(
+        INITIAL_AMOUNT
+          + MIN_RELAY_FEE_KICK_OFF_2
+          + DUST_AMOUNT // connector 3 to take 1
+          + MIN_RELAY_FEE_ASSERT_INITIAL
+          + MIN_RELAY_FEE_ASSERT_COMMIT1
+          + MIN_RELAY_FEE_ASSERT_COMMIT2
+          + MIN_RELAY_FEE_ASSERT_FINAL
+          + DUST_AMOUNT // connector 4 to take 2
+          + MIN_RELAY_FEE_DISPROVE,
```
Comment (aristotleee) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:121]: for simplicity, we can just give it an enough amount, we don't have to keep it precise unless we are testing the final output value
or we have a batch fee declared as `PEG_OUT_FEE`

File: bridge/tests/bridge/helper.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (franklynot) [reply bridge/tests/bridge/helper.rs:369]: Sorry, I mixed things up. You're right.

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 13]
```diff
@@ -133,8 +133,8 @@ pub async fn create_peg_out_graph() -> (BitVMClient, BitVMClient, String, Script
     faucet
         .fund_inputs(&verifier_0_operator_depositor, &funding_inputs)
         .await;
-    println!("Waiting for funding inputs tx...");
-    sleep(Duration::from_secs(TX_WAIT_TIME)).await;
+
+    wait_for_confirmation_with_message(config.network, Some("funding inputs")).await;
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:137]: In the case of a `Faucet` and awaiting funding inputs, there's a dedicated method `Faucet::wait()`. Just call it on the result of `fund_inputs()` above with no arguments. See how it's done across the test code (e.g. in `musig2_peg_in.rs:33`).

**Review by franklynot — CHANGES_REQUESTED** [review]

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 13 above)
Comment (franklynot) [reply bridge/tests/bridge/e2e/peg_out/utils.rs:137]: BTW, we should probably call that `Faucet::wait_for_funding_inputs()`. Feel free to rename.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (franklynot) [reply bridge/tests/bridge/e2e/peg_out/utils.rs:37]: If you mean the console output, it would still show every step. I was saying we'd print out the success message after the call.

[the remark this answers is not in the document]
Comment (franklynot) [reply bridge/tests/bridge/e2e/peg_out/utils.rs:37]: (Suggestion posted in a new comment.)

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+        .create_peg_in_graph(deposit_input, depositor_evm_address)
+        .await;
+
+    match verifier_0.broadcast_peg_in_deposit(&graph_id).await {
+        Ok(txid) => println!(
+            "Broadcasted {} with txid {txid}",
+            "peg-in deposit".bold().green()
+        ),
+        Err(e) => panic!("Failed to broadcast peg-in deposit: {e}"),
+    }
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs:43]: Is there a reason you want to avoid `panic` macro?

[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs:43]: `expect` does not return the error, that we print in the first scenario.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Refactored as discussed via call.

File: bridge/tests/bridge/setup.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/setup.rs]: Now we generate the proof just once.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 1 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Renamed to `depositor_verifier_0`.

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+            println!("Broadcasted {} with txid {txid}", "disprove".bold().red());
+            panic!("{}", "Incorrectly disproved correct ZK proof".bold().red());
+        }
+        Err(e) => println!(
+            "{}: {e}",
+            "Successfully rejected disproving correct ZK proof"
+                .bold()
+                .green()
+        ),
+    }
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/disprove_reject.rs]: Yes, let's use assert.

File: bridge/tests/bridge/e2e/peg_out/disprove_success.rs [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+            println!(
+                "{}",
+                "Succesfully disproved incorrect ZK proof".bold().green()
+            );
+        }
+        Err(e) => panic!(
+            "{}: {e}",
+            "Failed to disprove incorrect ZK proof".bold().red()
+        ),
+    }
```
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/disprove_success.rs]: Yes, let's use assert.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 12 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Refactored to use `PEG_OUT_FEE `.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 13 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Done

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Refactored as discussed via call.

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+    let result = verifier_1
+        .broadcast_disprove(&peg_out_graph_id, reward_script)
+        .await;
+
+    assert!(
+        matches!(result, Err(Error::Chunker(ChunkerError::ValidProof))),
+        "Should have failed with {} but got {:?}",
+        Error::Chunker(ChunkerError::ValidProof),
+        result
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/disprove_reject.rs:38]:
```suggestion
        result.unwrap_err()
```

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    .await;
+
+    let result = verifier_1
+        .broadcast_disprove(&peg_out_graph_id, reward_script)
+        .await;
+
+    assert!(
+        matches!(result, Err(Error::Chunker(ChunkerError::ValidProof))),
+        "Should have failed with {} but got {:?}",
+        Error::Chunker(ChunkerError::ValidProof),
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/disprove_reject.rs:37]: Perhaps this can be extracted into an `expected_error` variable and used in the `matches!` macro above as well?

File: bridge/src/client/cli/client_command.rs [hunk 19]
```diff
@@ -215,8 +216,16 @@ impl ClientCommand {
             Some(("kick_off_2", _)) => self.client.broadcast_kick_off_2(graph_id).await,
             Some(("start_time", _)) => self.client.broadcast_start_time(graph_id).await,
             Some(("assert_initial", _)) => self.client.broadcast_assert_initial(graph_id).await,
-            Some(("assert_commit_1", _)) => self.client.broadcast_assert_commit_1(graph_id).await,
-            Some(("assert_commit_2", _)) => self.client.broadcast_assert_commit_2(graph_id).await,
+            Some(("assert_commit_1", _)) => {
+                self.client
+                    .broadcast_assert_commit_1(graph_id, &get_proof())
```
Comment (franklynot) [inline bridge/src/client/cli/client_command.rs:221]: We also need to pass the proof in the call that creates the peg out graph. Then it must be forwarded to the connector C constructor, because it's needed there when calling `generate_assert_leaves()`.
It's actually interesting how the disprove tests are passing now, even though `generate_assert_leaves()` is working off of a different proof (i.e. the `RawProof::default()` instead of the one from `get_proof()`).

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+    graph_id
+}
+
+pub async fn create_peg_out_graph() -> (
+    BitVMClient,
+    BitVMClient,
+    String,
+    ScriptBuf,
+    Input,
+    Network,
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:86]: `Network` is redundant here. It's already returned in both clients (`BitVMClient::source_network`).

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+    faucet::{Faucet, FaucetType},
+    helper::{
+        generate_stub_outpoint, get_default_peg_out_event, print_tx_broadcasted,
+        wait_for_confirmation_with_message, wait_for_timelock_expiry,
+    },
+    setup::{setup_test, INITIAL_AMOUNT},
+};
+
+pub async fn create_peg_in_graph(
+    network: Network,
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:21]: `Network` redundant here as well (it's in the passed in clients).

File: bridge/tests/bridge/e2e/peg_out/utils.rs [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+            amount: peg_out_input_amount,
+        },
+        config.network,
+        config.valid_proof,
+        config.invalid_proof,
+    )
+}
+
+pub async fn broadcast_txs_for_disprove_scenario(
+    network: Network,
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/utils.rs:224]: Same here regarding the `network`.

File: bridge/tests/bridge/helper.rs [hunk 23]
```diff
@@ -301,16 +336,84 @@ pub fn get_lock_scripts_cached(
...
+        proof,
+        public: vec![c],
+        vk,
+    }
 }
 
-pub fn get_incorrect_proof() -> RawProof {
-    let mut correct_proof = get_correct_proof();
+pub fn invalidate_proof(valid_proof: &RawProof) -> RawProof {
+    let mut _valid_proof = valid_proof.clone();
```
Comment (franklynot) [inline bridge/tests/bridge/helper.rs:363]: [Minor] We should probably just call this var `invalid_proof` despite it being initially a clone of a valid one. It ends up being invalid is what matters. ;]

**Review by franklynot — CHANGES_REQUESTED** [review]

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    )
+    .await;
+
+    let result = verifier_1
+        .broadcast_disprove(&peg_out_graph_id, reward_script)
+        .await;
+
+    assert!(
+        matches!(result, Err(Error::Chunker(ChunkerError::ValidProof))),
+        "Should have failed with {} but got {:?}",
```
Comment (franklynot) [inline bridge/tests/bridge/e2e/peg_out/disprove_reject.rs:36]: Did you want to use colorize here as well?

File: bridge/tests/bridge/client/fee.rs [hunk 25]
```diff
@@ -32,7 +33,7 @@ use crate::bridge::{
     faucet::{Faucet, FaucetType},
     helper::{
         check_tx_output_sum, find_peg_in_graph_by_peg_out, generate_stub_outpoint,
-        get_incorrect_proof, get_reward_amount, wait_for_confirmation, wait_for_timelock_expiry,
+        get_reward_amount, get_valid_proof, wait_for_confirmation, wait_for_timelock_expiry,
```
Comment (franklynot) [inline bridge/tests/bridge/client/fee.rs:36]: Unused import: `get_valid_proof`.

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs (hunk 17 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/disprove_reject.rs:38]: We don't want to unwrap err, since in case of success, we want to panic here as well.

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs (hunk 18 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/disprove_reject.rs:37]: If we'd like to use extracted variable we would like to:
```
    let expected_error = Error::Chunker(ChunkerError::ValidProof);
    assert!(
        matches!(result, Err(expected_error))
    )
```
In this case, `expected_error` in the pattern is treated as a new variable, not the one defined outside and we get compilation error.

To use such variable we would have to:
```
    let expected_error = Error::Chunker(ChunkerError::ValidProof);
    assert!(
        matches!(result, Err(error) if error == expected_error)
    )
```
And we cannot compare 2 errors that way since the errors don't implement `Eq` nor `PartialEq` traits.

I suggest leaving it this way.

File: bridge/src/client/cli/client_command.rs (hunk 19 above)
Comment (StarDumpling) [reply bridge/src/client/cli/client_command.rs:221]: You're describing here `handle_broadcast_command()` fn that is not used in the e2e tests. This function is used only in cli.

Getting the proof for cli is a part of a `deliverable#2` task.

File: bridge/tests/bridge/client/fee.rs (hunk 25 above)
Comment (StarDumpling) [reply bridge/tests/bridge/client/fee.rs]: Done.

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs (hunk 24 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/disprove_reject.rs:34]: Yes, fixed.

File: bridge/tests/bridge/helper.rs (hunk 23 above)
Comment (StarDumpling) [reply bridge/tests/bridge/helper.rs]: Sure, done.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 21 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Done.

File: bridge/tests/bridge/e2e/peg_out/utils.rs (hunk 20 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/utils.rs]: Done.

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs [hunk 26]
```diff
@@ -0,0 +1,45 @@
+use bridge::error::{ChunkerError, Error};
+
+use colored::Colorize;
+
+use super::utils::{broadcast_txs_for_disprove_scenario, create_peg_out_graph};
+
+#[tokio::test]
+async fn test_e2e_disprove_reject() {
```
Comment (aristotleee) [inline bridge/tests/bridge/e2e/peg_out/disprove_reject.rs:8]: suggest to add #[serial] on e2e tests, they might interfere with each other

File: bridge/tests/bridge/e2e/peg_out/disprove_success.rs [hunk 27]
```diff
@@ -0,0 +1,41 @@
+use colored::Colorize;
+
+use super::utils::{broadcast_txs_for_disprove_scenario, create_peg_out_graph};
+
+#[tokio::test]
+async fn test_e2e_disprove_success() {
```
Comment (aristotleee) [inline bridge/tests/bridge/e2e/peg_out/disprove_success.rs:6]: suggest to add #[serial] on e2e tests, they might interfere with each other

**Review by aristotleee — APPROVED** [review]
I always had `PrecedingTxTimelockNotMet` error
changing `NUM_BLOCKS_REGTEST` in `src/utils.rs` to 3 worked for me

File: bridge/tests/bridge/e2e/peg_out/disprove_reject.rs (hunk 26 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/disprove_reject.rs:10]: Done

File: bridge/tests/bridge/e2e/peg_out/disprove_success.rs (hunk 27 above)
Comment (StarDumpling) [reply bridge/tests/bridge/e2e/peg_out/disprove_success.rs:8]: Done
